physics/module_SGSCloud_RadPre.F90

```diff
@@ -36,81 +36,105 @@ end subroutine sgscloud_radpre_finalize
            qr, qs, qg,                   &
-           qci_conv,                     &
+           qci_conv,ud_mf,               &
            imfdeepcnv, imfdeepcnv_gf,    &
-           qc_save, qi_save,             &
+           qc_save, qi_save, qs_save,    &
            qc_bl,qi_bl,cldfra_bl,        &
            delp,clouds1,clouds2,clouds3, &
-           clouds4,clouds5,slmsk,        &
+           clouds4,clouds5,              &
+           clouds8,clouds9,slmsk,        &
            nlay, plyr, xlat, dz,de_lgth, &
            cldsa,mtopa,mbota,            &
            imp_physics, imp_physics_gfdl,&
            iovr,                         &
            errmsg, errflg                )
 
-! should be moved to inside the mynn:
       use machine , only : kind_phys
       use module_radiation_clouds, only : gethml
-      use radcons, only: qmin               ! Minimum values for various calculations
-      use funcphys, only: fpvs              ! Function ot compute sat. vapor pressure over liq.
+      use radcons,  only: qmin          ! Minimum values for various calculations
+      use funcphys, only: fpvs          ! Function to compute sat. vapor pressure over liq.
 !------------------------------------------------------------------- 
       implicit none
 !------------------------------------------------------------------- 
       ! Interface variables
-      real(kind=kind_phys), intent(in) :: con_g, con_pi, eps, epsm1 
-      real (kind=kind_phys)         :: gfac
-      integer,          intent(in)  :: im, levs, imfdeepcnv, imfdeepcnv_gf, &
+      real(kind=kind_phys), intent(in) :: con_g, con_pi, eps, epsm1
+      real(kind=kind_phys), intent(in) :: r_v, cpv, rcp
+      real(kind=kind_phys), intent(in) :: xlv, xlf, cp
+      real(kind=kind_phys), intent(in) :: dt
+      real :: xls, xlvcp, xlscp !derived below
+      real(kind=kind_phys)             :: gfac
+      integer,             intent(in)  :: im, levs, imfdeepcnv, imfdeepcnv_gf, &
            &               nlay, imp_physics, imp_physics_gfdl
-      logical,          intent(in)  :: flag_init, flag_restart, do_mynnedmf
+      logical,             intent(in)  :: flag_init, flag_restart, do_mynnedmf
+
       real(kind=kind_phys), dimension(:,:), intent(inout) :: qc, qi
       real(kind=kind_phys), dimension(:,:), intent(inout) :: qr, qs, qg
       ! qci_conv only allocated if GF is used
-      real(kind=kind_phys), dimension(:,:),     intent(inout) :: qci_conv
-      real(kind=kind_phys), dimension(:,:), intent(in)    :: T3D,delp, &
-           &                                                     qv,P3D
-      real(kind=kind_phys), dimension(:,:), intent(inout) :: &
-           &         clouds1,clouds2,clouds3,clouds4,clouds5
-      real(kind=kind_phys), dimension(:,:), intent(inout) :: qc_save, qi_save
+      real(kind=kind_phys), dimension(:,:), intent(inout) :: qci_conv
+      real(kind=kind_phys), dimension(:,:), intent(in)    :: ud_mf
```

Comment:
On thing to be aware of: `ud_mf` is in the GFS_interstitial DDT and as far as I can see it gets set/reset in the physics group. Since it is intent(in) here, you probably get invalid data passed in (from a different block)

Reply:
You are wise to be concerned, but in this case, I think it's OK. Here is a brief printout of ud_mf, which has reasonable values of mass-flux (although sometime zero - maybe the convection isn't reaching that high at some i points):

 k=          20  ud_mf=  0.121096490473802     
 k=          20  ud_mf=  0.000000000000000E+000
 k=          20  ud_mf=  0.719252198946091     
 k=          20  ud_mf=   1.83385869435230     
 k=          20  ud_mf=   1.69616072660016     
 k=          32  ud_mf=   1.70287760640253     
 k=          32  ud_mf=   1.25508844531819     
 k=          32  ud_mf=   1.30862285882288     
 k=          32  ud_mf=   1.79306649029620     
 k=          32  ud_mf=  0.858321688490849     
 k=          32  ud_mf=  0.471099592930162     
 k=          32  ud_mf=  0.901912835814989     
 k=          32  ud_mf=   1.15600537456719     
 k=          20  ud_mf=   1.39565460657562     
 k=          20  ud_mf=   1.32165535435073     
 k=          21  ud_mf=  0.000000000000000E+000
 k=          21  ud_mf=  0.000000000000000E+000
 k=          21  ud_mf=  0.000000000000000E+000
 k=          21  ud_mf=  0.000000000000000E+000
 k=          21  ud_mf=  0.000000000000000E+000
 k=          21  ud_mf=  0.689175993672929     
 k=          21  ud_mf=  0.388614361705108 

And here is an example of the cloud fractions at the upper-levels produced by these modifications (right) compared to when leveraging the MYNN stratus cloud fraction (left):

<img width="995" alt="Screen Shot 2022-02-16 at 9 04 08 AM" src="https://user-images.githubusercontent.com/14096622/154306082-27e83cd5-3ffa-4415-a1b9-150b63cf6725.png">

These results are within the bounds of expectations to me.

Reply:
Just to be clear, I agree with Dom that this shouldn't work. I am investigating why it may **OR** how I can make sure it works.

Reply:
OK... I was puzzled as to how this code seems to work, so I:

1. added a "ud_mf_gf" in the TBD block,
2. set it ud_mf_gf(i,k) = ud_mf(i,k) in the GF driver,
3. piped ud_mf_gf into module_SGSCloud_RadPre.f90 instead of ud_mf.
4. Ended up with _almost_ identical results

I would happily attach an image for proof but this web interface is not allowing me to do so (error: "Something went really wrong, and we can't process that file).

So, even though either methods produce basically the same answer (although not bit-4-bit identical), I suggest we go with this updated method just because we know the TBD block is saved in memory. Note that I failed in converting ud_mf into a TBD array. It failed at compile time and I don't know why. Apparently, it takes more than simply modifying GFS_typedefs.F90 and GFS_debug.F90...
 
Happy to hear any comments on this.

Reply:
Thanks Dom. I missed the ud_mf move in GFS_typedefs.meta. Now it compiles and runs and I don't need the additional array. As far as I can tell, this code is ready for the regression tests.

Reply:
Sam,
Perhaps we can ask Haiqin and Hannah to review this PR?
-joe

Reply:
Just curious, did moving ud_mf into Tbd have an impact on results?

Reply:
No, by eye the cloud fractions looked identical, but it was not bit-4-bit identical.